clients/src/main/java/org/apache/kafka/clients/consumer/StubbedAsyncKafkaConsumer.java

```diff
@@ -0,0 +1,548 @@
+            return new ConsumerRecord<>(partition.topic(), partition.partition(), offset,
+                    timestamp, timestampType,
+                    keyByteArray == null ? ConsumerRecord.NULL_SIZE : keyByteArray.length,
+                    valueByteArray == null ? ConsumerRecord.NULL_SIZE : valueByteArray.length,
+                    key, value, headers, leaderEpoch);
+        } catch (RuntimeException e) {
+            throw new RecordDeserializationException(partition, record.offset(),
+                    "Error deserializing key/value for partition " + partition +
+                            " at offset " + record.offset() + ". If needed, please seek past the record to continue consumption.", e);
+        }
+    }
+
+    @Override
+    public void commitSync() {
+        commitSync(Duration.ofMillis(defaultApiTimeoutMs));
+    }
+
+    @Override
+    public void commitSync(Duration timeout) {
+        commitSync(subscriptions.allConsumed(), timeout);
+    }
+
+    @Override
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        commitSync(offsets, Duration.ofMillis(defaultApiTimeoutMs));
+    }
+
+    @Override
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration timeout) {
+        // The background thread will do the following when it receives this event:
+        //
+        // 1. Metadata will update its 'last seen' epoch, if newer
+        // 2. Invoke the coordinator commitOffsetsSync logic
+        eventHandler.addAndGet(new CommitSyncEvent(offsets), time.timer(timeout));
+    }
+
+    @Override
+    public void commitAsync() {
+        commitAsync(null);
+    }
+
+    @Override
+    public void commitAsync(OffsetCommitCallback callback) {
+        commitAsync(subscriptions.allConsumed(), callback);
+    }
+
+    @Override
+    public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
+        // The background thread will do the following when it receives this event:
+        //
+        // 1. Metadata will update its 'last seen' epoch, if newer
+        // 2. Invoke the coordinator commitOffsetsAsync logic
+        eventHandler.add(new CommitAsyncEvent(offsets, callback));
+    }
+
+    @Override
+    public void seek(TopicPartition partition, long offset) {
+        // TODO: This needs to be propagated to the background thread, right?
+        subscriptions.seek(partition, offset);
+    }
+
+    @Override
+    public void seek(TopicPartition partition, OffsetAndMetadata offsetAndMetadata) {
+        // TODO: This needs to be propagated to the background thread, right?
+    }
+
+    @Override
+    public void seekToBeginning(Collection<TopicPartition> partitions) {
+
+    }
+
+    @Override
+    public void seekToEnd(Collection<TopicPartition> partitions) {
+
+    }
+
+    @Override
+    public long position(TopicPartition partition) {
```

Comment:
Again I'd like to emphasize the ordering of this function with other func here. And just to demo here's an example:

```
0: buffered records from offset 0 to offset 10 (exclusive), no acked position offset, fetch offset 10 (i.e. for the next fetch request).
1: Foreground **poll** returned records [0, 5). **Buffered** [5, 10), acked **position** offset 0, **fetch** offset 10.
2: Background gets records [10, 20) from fetch response. **Buffered** [5, 20), acked **position** offset 0, **fetch** offset 20.
3: Foreground **position** call returns 0 still.
4: Foreground **poll** returned records [5, 15). **Buffered** [15, 20), acked **position** offset 5, **fetch** offset 20.
5: Foreground **position** call returns 5.
6: Foreground **seekToBeginning**. **Buffered** [15, 20), acked **position** offset 0, **fetch** offset 20.
7: Foreground **position** call returns 0.
8: Foreground **poll** calls again, here cannot return any data yet since the buffered records does not contains any for the starting position.
9: Background clears the buffered records, reset its **fetch** offset to 0 and send the next fetch request.
10: Background finally gets some data from the fetch response, say [0, 15), before this is done all **poll** calls from foreground should return nothing.
11: The next **poll** call could then return records [0, 10), acked position is still 0, and fetch offset is 15.
```